### DISCUSSION TITLE
単体テストの考え方/使い方 5章

### DISCUSSION THREAD

**Discussion by dameleon** [atman-inc/training#100] [General]
輪読会のサマリ用ディスカッション。
5章分の各章単位程度でコメントを記載するので、不明点、相談点、確認点、勉強になったポイントなどは、各自ディスカッションのスレッドへコメントをお願いします。

**Comment by dameleon** [comment]
## 5.3 モックの利用とテストの壊れやすさとの関係
### 5.3.1 ヘキサゴナルアーキテクチャとは？

前回より: https://github.com/atman-inc/training/discussions/99?sort=new#discussioncomment-5748683

**Reply by nmurakami0** [reply]
感想

- 気をつけないと、DB読み込みや外部システムの呼び出しを、ドメイン層でやってしまいそう
    - アプリケーション・サービス層が正解
    - ただ、5.4.1を見るとアクセス対象がテスト対象アプリケーションに閉じていれば、ドメイン層でDB読み込みしても良い？

**Reply by k-itachida** [reply]
感想
(ここら辺が話が難しくなってくる...)
- 概念としては非常にわかりやすい。Reactでの関心の分離と考え方は似ている（関数型プログラミングの話もこのあと出てくる）。
- 実際に実装した場合、きっちり境界線を引くのが難しそうに感じる。

**Reply by koushisa** [reply]
Thinking in  Redux

https://slides.com/jenyaterpil/redux-from-twitter-hype-to-production#/27

- アプリケーション = what = dispatch
- ドメイン = how = reducer

依存関係をdispatchからの一方通行にする

**Comment by dameleon** [comment]
### 5.3.2 システム内コミュニケーションとシステム間コミュニケーション
### 5.3.3 システム内コミュニケーションとシステム間コミュニケーション：例による比較

**Reply by y-kaneda** [reply]
- システム内コミュニケーション
  - 実装の詳細
  - 観察可能な振る舞いの一部にはならない
    - プライベートな実装ということ？
- システム間コミュニケーション
  - 外部サービス・外部システム
  - 観察可能な振る舞い
    - インターフェースは満たさなければならないという話っぽい？
  - テスト対象のプロダクションコードの実装を変えても影響がないので mock に置き換え可能

**Reply by nmurakami0** [reply]
- 異なるアプリケーション(システム間コミュニケーションが存在)がともに進化する上で、後方互換性を維持することが重要
    - 検証するのにモックが重要
    - システム内コミュニケーションでモックを使ってしまうと、テストが実装の詳細に深く結びついてしまい、リファクタリングへの耐性を失う
- リスト5.11
  - モックを使うがために、テストケースが実装の詳細と結びついてしまっているので、リファクタリングへの耐性が失われている

**Comment by dameleon** [comment]
## 5.4 振り返り：単体テストの古典学派とロンドン学派の違い

**Reply by y-kaneda** [reply]
- ロンドン学派は何でもかんでもモックにしてしまうがために、内部実装間の関係性に関してリファクタリングへの耐性が損なわれる

**Comment by dameleon** [comment]
~ 2023/05/12はここまで ~

**Comment by dameleon** [comment]
### 5.4.1 モックに置き換える必要のないプロセス外依存

**Reply by nmurakami0** [reply]
- 外部から観察できないプロセス外依存とのコミュニケーションは実装の詳細となる
    - e.g., データベース
        - データベースに対してテスト対象のアプリケーションを除く全てのアプリケーションからアクセスされることがないのであれば、テスト対象のアプリケーションとデータベース間のコミュニケーションを自由に変えられる
- モックは振る舞いを検証する場合もあるが、常にそうではない
    - アプリケーションの協会を超えて行われるコミュニケーションで、かつ、そのコミュニケーションによって生じる副作用を外部から見ることが出来る場合のみ

**Reply by y-kaneda** [reply]
- (本書での古典学派でのデータベースのモック化の話は共通で固定のDBの前提)
- 例えばinfra層のテストとかはプロセス外依存(DB)を含めての内部実装という話？
- infra層のような処理を置いておけばプロセス外依存を外部から意識しなくて良くなるというような話？

**Comment by dameleon** [comment]
### 5.4.2 モックを使った振る舞いの検証
